app/vlinsert/loki/loki_json.go

```diff
@@ -0,0 +1,133 @@
+package loki
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/valyala/fastjson"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/bytesutil"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logstorage"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/protoparser/common"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/writeconcurrencylimiter"
+
+	"github.com/VictoriaMetrics/metrics"
+)
+
+var (
+	rowsIngestedTotalJSON = metrics.NewCounter(`vl_rows_ingested_total{type="loki", format="json"}`)
+	parserPool            fastjson.ParserPool
+)
+
+func handleJSON(r *http.Request, w http.ResponseWriter) bool {
+	contentType := r.Header.Get("Content-Type")
+	reader := r.Body
+	if contentType == "gzip" {
+		zr, err := common.GetGzipReader(reader)
+		if err != nil {
+			httpserver.Errorf(w, r, "cannot read gzipped request: %s", err)
+			return true
+		}
+		defer common.PutGzipReader(zr)
+		reader = zr
+	}
+
+	cp, err := getCommonParams(r)
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot parse request: %s", err)
+		return true
+	}
+	lr := logstorage.GetLogRows(cp.StreamFields, cp.IgnoreFields)
+
+	processLogMessage := cp.GetProcessLogMessageFunc(lr)
+	n, err := processJSONRequest(reader, processLogMessage)
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot decode loki request: %s", err)
+		return true
+	}
+	rowsIngestedTotalJSON.Add(n)
+	return true
+}
+
+func processJSONRequest(r io.Reader, processLogMessage func(timestamp int64, fields []logstorage.Field)) (int, error) {
+	wcr := writeconcurrencylimiter.GetReader(r)
+	defer writeconcurrencylimiter.PutReader(wcr)
+
+	bytes, err := io.ReadAll(wcr)
+	if err != nil {
+		return 0, fmt.Errorf("cannot read request body: %w", err)
+	}
+
+	p := parserPool.Get()
+	defer parserPool.Put(p)
+	v, err := p.ParseBytes(bytes)
+	if err != nil {
+		return 0, fmt.Errorf("cannot parse request body: %w", err)
+	}
+
+	rowsIngested := 0
+	for stIdx, st := range v.GetArray("streams") {
+		// `stream` contains labels for the stream.
+		// Labels are same for all entries in the stream.
+		logFields := st.GetObject("stream")
+		if logFields == nil {
+			logger.Warnf("missing streams field from %q", st)
+			logFields = &fastjson.Object{}
+		}
+		commonFields := make([]logstorage.Field, logFields.Len()+1)
```

Comment:
Could you please elaborate on this one a bit?

Currently, common fields will be reused for all log entries for the same stream because fields are defined on the stream level.
For the next stream a new set of common fields will be created. Reusing previous slice might be problematic for cases when streams have different sets of labels(which is a quite common case).

Reply:
There is a slice trick:

```go
func ResizeNoCopyMayOverallocate[T any](dst []T, n int) []T {
	if n <= cap(dst) {
		return dst[:n]
	}
        nNew := roundToNearestPow2(n)
        dstNew := make([]T,nNew)
	return dst[:n]
}

func roundToNearestPow2(n int) int {
	pow2 := uint8(bits.Len(uint(n - 1)))
	return 1 << pow2
}
```

usage:

```go
commonFields := ResizeNoCopyMayOverallocate(commonFields, logFields.Len()+1)
```

 Currently creation of commonFields allocates memory.